Description: Build system fixes
 1. Strip -Werror and -pedantic.
 2. Fix gcrypt discovery.
 3. Explicitly discover dl.
 4. Explicitly add some more libraries.
 5. Fix LDFLAGS vs. LDADD to allow -Wl,--as-needed.
 6. Clean up linking of shared libraries.
Author: Dominik George <nik@naturalnet.de>
diff -ur a/configure.ac b/configure.ac
--- a/configure.ac	2018-12-21 22:17:27.000000000 +0100
+++ b/configure.ac	2019-02-01 15:51:23.485782297 +0100
@@ -19,7 +19,7 @@
 
 AC_PREREQ([2.61])
 AC_INIT([guacamole-server], [1.0.0])
-AM_INIT_AUTOMAKE([-Wall -Werror foreign subdir-objects])
+AM_INIT_AUTOMAKE([-Wall foreign subdir-objects])
 AM_SILENT_RULES([yes])
 
 LT_PREREQ([2.2])
@@ -534,9 +534,9 @@
 
     # libfreerdp-core / libfreerdp
     AC_CHECK_LIB([freerdp-core], [freerdp_new],
-                 [RDP_LIBS="$RDP_LIBS -lfreerdp-core"],
+                 [RDP_LIBS="$RDP_LIBS -lfreerdp-core -lfreerdp-common -lfreerdp-client -lwinpr-utils"],
                  [AC_CHECK_LIB([freerdp], [freerdp_new],
-                               [RDP_LIBS="$RDP_LIBS -lfreerdp -lfreerdp-client"],
+                               [RDP_LIBS="$RDP_LIBS -lfreerdp -lfreerdp-common -lfreerdp-client -lwinpr-utils"],
                                [AC_MSG_WARN([
   --------------------------------------------
    Unable to find libfreerdp-core / libfreerdp
@@ -1073,13 +1073,14 @@
     AC_CHECK_LIB([ssh2], [gcry_control],
                          [AC_CHECK_HEADER(gcrypt.h,
                                           [AC_DEFINE([LIBSSH2_USES_GCRYPT],,
-                                                     [Whether libssh2 was built against libgcrypt])],
+                                                     [Whether libssh2 was built against libgcrypt])
+                                           SSH_LIBS="$SSH_LIBS -lgcrypt"],
                                           [AC_MSG_WARN([
   --------------------------------------------
    libssh2 appears to be built against libgcrypt, but the libgcrypt headers
    could not be found. SSH will be disabled.
   --------------------------------------------])
-                                           have_libssh2=no])])
+                                           have_libssh2=no])], [], [-lgcrypt])
 
 fi
 
diff -ur a/src/common/Makefile.am b/src/common/Makefile.am
--- a/src/common/Makefile.am	2018-12-21 22:17:27.000000000 +0100
+++ b/src/common/Makefile.am	2019-02-01 15:51:23.485782297 +0100
@@ -57,7 +57,7 @@
     surface.c
 
 libguac_common_la_CFLAGS =  \
-    -Werror -Wall -pedantic \
+    -Wall                   \
     @LIBGUAC_INCLUDE@
 
 libguac_common_la_LIBADD = \
diff -ur a/src/common-ssh/Makefile.am b/src/common-ssh/Makefile.am
--- a/src/common-ssh/Makefile.am	2018-12-21 22:17:27.000000000 +0100
+++ b/src/common-ssh/Makefile.am	2019-02-01 15:51:23.486782297 +0100
@@ -41,14 +41,12 @@
     common-ssh/user.h
 
 libguac_common_ssh_la_CFLAGS = \
-    -Werror -Wall -pedantic    \
+    -Wall                      \
     @COMMON_INCLUDE@           \
     @LIBGUAC_INCLUDE@
 
 libguac_common_ssh_la_LIBADD = \
-    @LIBGUAC_LTLIB@
-
-libguac_common_ssh_la_LDFLAGS = \
+    @LIBGUAC_LTLIB@             \
     @PTHREAD_LIBS@              \
     @SSH_LIBS@                  \
     @SSL_LIBS@
diff -ur a/src/guacd/Makefile.am b/src/guacd/Makefile.am
--- a/src/guacd/Makefile.am	2018-12-21 22:17:27.000000000 +0100
+++ b/src/guacd/Makefile.am	2019-02-01 15:51:23.486782297 +0100
@@ -48,16 +48,14 @@
     proc-map.c
 
 guacd_CFLAGS =              \
-    -Werror -Wall -pedantic \
+    -Wall                   \
     @COMMON_INCLUDE@        \
     @LIBGUAC_INCLUDE@
 
 guacd_LDADD =        \
     @COMMON_LTLIB@   \
-    @LIBGUAC_LTLIB@
-
-guacd_LDFLAGS =    \
-    @PTHREAD_LIBS@ \
+    @LIBGUAC_LTLIB@  \
+    @PTHREAD_LIBS@   \
     @SSL_LIBS@
 
 EXTRA_DIST =            \
diff -ur a/src/libguac/Makefile.am b/src/libguac/Makefile.am
--- a/src/libguac/Makefile.am	2018-12-21 22:17:27.000000000 +0100
+++ b/src/libguac/Makefile.am	2019-02-01 15:51:23.486782297 +0100
@@ -115,11 +115,13 @@
 endif
 
 libguac_la_CFLAGS = \
-    -Werror -Wall -pedantic -I$(srcdir)/guacamole
+    -Wall -I$(srcdir)/guacamole
 
 libguac_la_LDFLAGS =     \
     -version-info 16:0:0 \
-    -no-undefined        \
+    -no-undefined
+
+libguac_la_LIBADD = \
     @CAIRO_LIBS@         \
     @DL_LIBS@            \
     @JPEG_LIBS@          \
diff -ur a/src/protocols/rdp/Makefile.am b/src/protocols/rdp/Makefile.am
--- a/src/protocols/rdp/Makefile.am	2018-12-21 22:17:27.000000000 +0100
+++ b/src/protocols/rdp/Makefile.am	2019-02-01 15:51:23.487782296 +0100
@@ -140,96 +140,100 @@
 #
 
 libguac_client_rdp_la_CFLAGS = \
-    -Werror -Wall -Iinclude    \
+    -Wall -Iinclude            \
     @COMMON_INCLUDE@           \
     @COMMON_SSH_INCLUDE@       \
     @LIBGUAC_INCLUDE@
 
 libguac_client_rdp_la_LDFLAGS = \
-    -version-info 0:0:0         \
-    @CAIRO_LIBS@                \
-    @PTHREAD_LIBS@              \
-    @RDP_LIBS@
+    -version-info 0:0:0
 
-libguac_client_rdp_la_LIBADD =     \
-    @COMMON_LTLIB@                 \
-    @LIBGUAC_LTLIB@
+libguac_client_rdp_la_LIBADD = \
+    @COMMON_LTLIB@             \
+    @LIBGUAC_LTLIB@            \
+    @CAIRO_LIBS@               \
+    @PTHREAD_LIBS@             \
+    @RDP_LIBS@
 
 #
 # RDPDR
 #
 
 guacdr_cflags                = \
-    -Werror -Wall -Iinclude    \
+    -Wall -Iinclude            \
     @COMMON_INCLUDE@           \
     @COMMON_SSH_INCLUDE@       \
     @LIBGUAC_INCLUDE@
 
 guacdr_ldflags =                   \
-    -module -avoid-version -shared \
-    @PTHREAD_LIBS@                 \
-    @RDP_LIBS@
+    -module -avoid-version -shared
 
 guacdr_libadd =     \
     @COMMON_LTLIB@  \
-    @LIBGUAC_LTLIB@
+    @LIBGUAC_LTLIB@ \
+    @PTHREAD_LIBS@  \
+    @RDP_LIBS@      \
+    libguac-client-rdp.la
 
 #
 # Audio Input
 #
 
 guacai_cflags               =  \
-    -Werror -Wall -Iinclude    \
+    -Wall -Iinclude            \
     @COMMON_INCLUDE@           \
     @COMMON_SSH_INCLUDE@       \
     @LIBGUAC_INCLUDE@
 
 guacai_ldflags =                   \
-    -module -avoid-version -shared \
-    @PTHREAD_LIBS@                 \
-    @RDP_LIBS@
+    -module -avoid-version -shared
 
 guacai_libadd =     \
     @COMMON_LTLIB@  \
-    @LIBGUAC_LTLIB@
+    @LIBGUAC_LTLIB@ \
+    @PTHREAD_LIBS@  \
+    @RDP_LIBS@      \
+    libguac-client-rdp.la
 
 #
 # RDPSND
 #
 
 guacsnd_cflags               = \
-    -Werror -Wall -Iinclude    \
+    -Wall -Iinclude            \
     @COMMON_INCLUDE@           \
     @COMMON_SSH_INCLUDE@       \
     @LIBGUAC_INCLUDE@
 
 guacsnd_ldflags =                  \
-    -module -avoid-version -shared \
-    @PTHREAD_LIBS@                 \
-    @RDP_LIBS@
+    -module -avoid-version -shared
 
 guacsnd_libadd =    \
     @COMMON_LTLIB@  \
-    @LIBGUAC_LTLIB@
+    @LIBGUAC_LTLIB@ \
+    @PTHREAD_LIBS@  \
+    @RDP_LIBS@      \
+    libguac-client-rdp.la
 
 #
 # Static Virtual Channels
 #
 
 guacsvc_cflags               = \
-    -Werror -Wall -Iinclude    \
+    -Wall -Iinclude            \
     @COMMON_INCLUDE@           \
     @COMMON_SSH_INCLUDE@       \
     @LIBGUAC_INCLUDE@
 
 guacsvc_ldflags =                  \
-    -module -avoid-version -shared \
-    @PTHREAD_LIBS@                 \
-    @RDP_LIBS@
+    -module -avoid-version -shared
 
 guacsvc_libadd =    \
     @COMMON_LTLIB@  \
-    @LIBGUAC_LTLIB@
+    @LIBGUAC_LTLIB@ \
+    @PTHREAD_LIBS@  \
+    @RDP_LIBS@      \
+    libguac-client-rdp.la
 
 #
 # Optional SFTP support
diff -ur a/src/protocols/ssh/Makefile.am b/src/protocols/ssh/Makefile.am
--- a/src/protocols/ssh/Makefile.am	2018-12-21 22:17:27.000000000 +0100
+++ b/src/protocols/ssh/Makefile.am	2019-02-01 15:51:23.487782296 +0100
@@ -49,7 +49,7 @@
 endif
 
 libguac_client_ssh_la_CFLAGS = \
-    -Werror -Wall -Iinclude    \
+    -Wall -Iinclude            \
     @COMMON_SSH_INCLUDE@       \
     @LIBGUAC_INCLUDE@          \
     @TERMINAL_INCLUDE@
@@ -58,11 +58,11 @@
     @COMMON_LTLIB@             \
     @COMMON_SSH_LTLIB@         \
     @LIBGUAC_LTLIB@            \
-    @TERMINAL_LTLIB@
+    @TERMINAL_LTLIB@           \
+    @PTHREAD_LIBS@             \
+    @SSH_LIBS@                 \
+    @SSL_LIBS@
 
 libguac_client_ssh_la_LDFLAGS = \
-    -version-info 0:0:0         \
-    @PTHREAD_LIBS@              \
-    @SSH_LIBS@                  \
-    @SSL_LIBS@
+    -version-info 0:0:0
 
diff -ur a/src/protocols/telnet/Makefile.am b/src/protocols/telnet/Makefile.am
--- a/src/protocols/telnet/Makefile.am	2018-12-21 22:17:27.000000000 +0100
+++ b/src/protocols/telnet/Makefile.am	2019-02-01 15:51:23.488782296 +0100
@@ -39,17 +39,17 @@
     user.h
 
 libguac_client_telnet_la_CFLAGS = \
-    -Werror -Wall -Iinclude       \
+    -Wall -Iinclude               \
     @LIBGUAC_INCLUDE@             \
     @TERMINAL_INCLUDE@
 
 libguac_client_telnet_la_LIBADD = \
     @COMMON_LTLIB@                \
     @LIBGUAC_LTLIB@               \
-    @TERMINAL_LTLIB@
+    @TERMINAL_LTLIB@              \
+    @PTHREAD_LIBS@                \
+    @TELNET_LIBS@
 
 libguac_client_telnet_la_LDFLAGS = \
-    -version-info 0:0:0            \
-    @PTHREAD_LIBS@                 \
-    @TELNET_LIBS@
+    -version-info 0:0:0
 
diff -ur a/src/protocols/vnc/Makefile.am b/src/protocols/vnc/Makefile.am
--- a/src/protocols/vnc/Makefile.am	2018-12-21 22:17:27.000000000 +0100
+++ b/src/protocols/vnc/Makefile.am	2019-02-01 15:51:23.488782296 +0100
@@ -47,20 +47,20 @@
     vnc.h
 
 libguac_client_vnc_la_CFLAGS =        \
-    -Werror -Wall -pedantic -Iinclude \
+    -Wall -Iinclude                   \
     @COMMON_INCLUDE@                  \
     @COMMON_SSH_INCLUDE@              \
     @LIBGUAC_INCLUDE@                 \
     @PULSE_INCLUDE@
 
 libguac_client_vnc_la_LDFLAGS = \
-    -version-info 0:0:0         \
-    @CAIRO_LIBS@                \
-    @VNC_LIBS@ 
+    -version-info 0:0:0
 
 libguac_client_vnc_la_LIBADD = \
     @COMMON_LTLIB@             \
-    @LIBGUAC_LTLIB@
+    @LIBGUAC_LTLIB@            \
+    @CAIRO_LIBS@               \
+    @VNC_LIBS@
 
 # Optional SFTP support
 if ENABLE_COMMON_SSH
diff -ur a/src/pulse/Makefile.am b/src/pulse/Makefile.am
--- a/src/pulse/Makefile.am	2018-12-21 22:17:27.000000000 +0100
+++ b/src/pulse/Makefile.am	2019-02-01 15:51:23.488782296 +0100
@@ -29,12 +29,10 @@
     pulse.c
 
 libguac_pulse_la_CFLAGS =  \
-    -Werror -Wall -pedantic \
+    -Wall                  \
     @LIBGUAC_INCLUDE@
 
 libguac_pulse_la_LIBADD = \
-    @LIBGUAC_LTLIB@
-
-libguac_pulse_la_LDFLAGS = \
+    @LIBGUAC_LTLIB@       \
     @PULSE_LIBS@
 
diff -ur a/src/terminal/Makefile.am b/src/terminal/Makefile.am
--- a/src/terminal/Makefile.am	2018-12-21 22:17:27.000000000 +0100
+++ b/src/terminal/Makefile.am	2019-02-01 15:51:23.489782295 +0100
@@ -50,16 +50,14 @@
     xparsecolor.c
 
 libguac_terminal_la_CFLAGS = \
-    -Werror -Wall            \
+    -Wall                    \
     @COMMON_INCLUDE@         \
     @LIBGUAC_INCLUDE@        \
     @PANGO_CFLAGS@           \
     @PANGOCAIRO_CFLAGS@
 
 libguac_terminal_la_LIBADD = \
-    @LIBGUAC_LTLIB@
-
-libguac_terminal_la_LDFLAGS = \
+    @LIBGUAC_LTLIB@          \
     @CAIRO_LIBS@              \
     @MATH_LIBS@               \
     @PANGO_LIBS@              \
diff -ur a/tests/Makefile.am b/tests/Makefile.am
--- a/tests/Makefile.am	2018-12-21 22:17:27.000000000 +0100
+++ b/tests/Makefile.am	2019-02-01 15:51:23.489782295 +0100
@@ -49,7 +49,7 @@
     util/guac_unicode.c
 
 test_libguac_CFLAGS =       \
-    -Werror -Wall -pedantic \
+    -Wall                   \
     @COMMON_INCLUDE@        \
     @LIBGUAC_INCLUDE@
 
